Discovery: ConcurrentModificationException in Bucket.getLiveNodes

### Description
```
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:936)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at java.base/java.util.concurrent.CompletableFuture$Timeout.run(CompletableFuture.java:2874)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1604)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.ethereum.beacon.discovery.storage.KBucket.getLiveNodes(KBucket.java:52)
	at org.ethereum.beacon.discovery.storage.KBuckets.lambda$getLiveNodeRecords$0(KBuckets.java:50)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
	at java.base/java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:411)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.ethereum.beacon.discovery.storage.KBucketsIterator.updateCurrentBatch(KBucketsIterator.java:61)
	at org.ethereum.beacon.discovery.storage.KBucketsIterator.hasNext(KBucketsIterator.java:40)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at tech.pegasys.teku.networking.p2p.connection.ConnectionManager.lambda$connectToBestPeers$1(ConnectionManager.java:114)
	at tech.pegasys.teku.networking.eth2.peers.Eth2PeerSelectionStrategy.selectPeersToConnect(Eth2PeerSelectionStrategy.java:86)
	at tech.pegasys.teku.networking.p2p.connection.ConnectionManager.connectToBestPeers(ConnectionManager.java:107)
	at tech.pegasys.teku.networking.p2p.connection.ConnectionManager.lambda$searchForPeers$2(ConnectionManager.java:131)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$finish$22(SafeFuture.java:304)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	... 10 more
```

--- a/src/main/java/org/ethereum/beacon/discovery/storage/KBuckets.java
+++ b/src/main/java/org/ethereum/beacon/discovery/storage/KBuckets.java
@@ -10,6 +10,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Spliterator;
 import java.util.Spliterators;
+import java.util.function.Function;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import org.apache.tuweni.bytes.Bytes;
@@ -47,14 +48,36 @@ public class KBuckets {
     if (distance == 0) {
       return Stream.of(localNodeRecordStore.getLocalNodeRecord());
     }
-    return getBucket(distance).stream().flatMap(bucket -> bucket.getLiveNodes().stream());
+    return streamFromBucket(distance, bucket -> bucket.getLiveNodes().stream());
   }
 
   public synchronized Stream<NodeRecord> getAllNodeRecords(int distance) {
     if (distance == 0) {
       return Stream.of(localNodeRecordStore.getLocalNodeRecord());
     }
-    return getBucket(distance).stream().flatMap(bucket -> bucket.getAllNodes().stream());
+    return streamFromBucket(distance, bucket -> bucket.getAllNodes().stream());
+  }
+
+  /**
+   * Extracts a {@link Stream} from a {@link KBucket} at a given distance, ensuring that all
+   * accesses to the KBucket are completed prior to the method returning.
+   *
+   * <p>Specifically this method avoids using {@code getBucket(distance).stream().flatMap(mapper)}
+   * as the mapper is then called after the method returns. That leads to thread safety issues.
+   *
+   * @param distance the bucket distance
+   * @param mapper the function to convert the KBucket to a Stream of objects
+   * @param <T> the type of object in the returned Stream
+   * @return the result of applying mapper to the bucket at the requested distance, or an empty
+   *     stream if there is no bucket at that distance.
+   */
+  private <T> Stream<T> streamFromBucket(int distance, final Function<KBucket, Stream<T>> mapper) {
+    final Optional<KBucket> bucket = getBucket(distance);
+    if (bucket.isPresent()) {
+      return mapper.apply(bucket.get());
+    } else {
+      return Stream.empty();
+    }
   }
 
   public Stream<NodeRecord> streamClosestNodes(Bytes nodeId) {

